cli/src/main.rs

```diff
@@ -0,0 +1,99 @@
+use std::{io::Write, path::PathBuf};
+
+use clap::Parser;
+use color_eyre::Result;
+
+mod path;
+
+#[derive(Parser, Debug)]
+struct Args {
+    /// Defines where your TS bindings will be saved by setting TS_RS_EXPORT_DIR
+    #[arg(long, short, default_value = "./bindings")]
+    output_directory: PathBuf,
+
+    /// Disables warnings caused by using serde attributes that ts-rs cannot process
+    #[arg(long)]
+    no_warnings: bool,
+
+    /// Adds the ".js" extension to import paths
+    #[arg(long)]
+    esm_imports: bool,
+
+    /// Formats the generated TypeScript files
+    #[arg(long)]
+    format: bool,
+
+    #[arg(long = "index")]
+    generate_index_ts: bool,
+}
+
+macro_rules! feature {
+    ($cargo_invocation: expr, $args: expr, { $($field: ident => $feature: literal),* $(,)? }) => {
+        $(
+            if $args.$field {
+                $cargo_invocation
+                    .arg("--features")
+                    .arg(format!("ts-rs/{}", $feature));
+            }
+        )*
+    };
+}
+
+fn main() -> Result<()> {
+    color_eyre::install()?;
+
+    let args = Args::parse();
+
+    let mut cargo_invocation = std::process::Command::new("cargo");
+
+    let metadata_path = args.output_directory.join("ts_rs.meta");
+    if metadata_path.exists() {
+        std::fs::remove_file(&metadata_path)?;
+    }
+
+    cargo_invocation
+        .arg("test")
+        .arg("export_bindings_")
+        .arg("--features")
+        .arg("ts-rs/export")
+        .env("TS_RS_EXPORT_DIR", path::absolute(&args.output_directory)?);
+
+    feature!(cargo_invocation, args, {
+        no_warnings => "no-serde-warnings",
+        esm_imports => "import-esm",
+        format => "format",
+    });
+
+    cargo_invocation.spawn()?.wait()?;
+
+    if args.generate_index_ts {
+        let metadata = std::fs::read_to_string(&metadata_path)?
+            .lines()
+            .map(ToOwned::to_owned)
+            .collect::<std::collections::HashSet<_>>();
+
+        let index_path = args.output_directory.join("index.ts");
+
+        if index_path.exists() {
+            std::fs::remove_file(&index_path)?;
+        }
+
+        if !metadata.is_empty() {
+            let mut index = std::fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(index_path)?;
+
+            index.write_all(
+                b"// This file was generated by [ts-rs](https://github.com/Aleph-Alpha/ts-rs). Do not edit this file manually.\n\n"
+            )?;
+            for file in metadata.iter() {
+                index.write_fmt(format_args!("export * from {file:?};\n"))?;
```

Comment:
Ah, that's cool! Thought we had to name the types individually.  
As it turns out, since [TS 5.0](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#support-for-export-type), there's also `export type * from ..`, which we probably want to use here.

Reply:
Hm, I guess we still need to name them individually, to avoid name collisions.  
In these cases, we'd want to emit something like `export type { A as A2, B } from "./a.ts";`

Reply:
We could also leave it to the user to resolve naming collisions by using different names for their types or using `#[ts(rename)]`. I think that if we rename the type it might be too confusing (i.e. user's type will be different than what they specified in Rust) and complex (i.e. we'd have to make sure that, if there are two types named `A`, we **always** rename the same one to `A1` and the other to `A2`, otherwise running the CLI twice in a row can change which type is which)

Reply:
Good point!

We'd want to disambiguate the names of both types, but I don't see a nice way of doing that. Even if we could, then adding a type later could cause a name conflict, which would break usages of the type which was already there.

So we could either detect the conflicts & show a nice error, or just generate invalid TS and let `tsc` tell the user what happened.

Reply:
> So we could either detect the conflicts

That would be awesome if we managed to pull it off

Reply:
I've got an idea, lemme see if that'd work out..

Reply:
> That would be awesome if we managed to pull it 

The metadata file could have both the type's name as well as its path, separated by a `;` or `,` This would basically make it a csv file

If this is done, it's important that the type name comes first, as it is easier to use character like `,`/`;` to separate them as those are not valid for TS type names, and therefore can be safely assumed to not be part of the type name. That's harder to find with file paths

Reply:
Then, in the CLI, we just check if any type name shows up more than once with a `HashMap<PathBuf, Vec<String>>` so we can say which types are repeated and to what files they were exported

Reply:
Oh no, you were faster :D

Reply:
I got it to work, might need some cleanup though

![image](https://github.com/Aleph-Alpha/ts-rs/assets/131818645/ca518816-5318-47b9-a498-de2cd048bfae)

Reply:
Nice! Might be cool to also have the `std::any::type_name` (or is there something better?) in the metadata file.  
Then, we could report the module path of the types as well.

Reply:
`color_eyre` also reexports a trait that allows for coloring text:
![image](https://github.com/Aleph-Alpha/ts-rs/assets/131818645/e6b33f78-56c6-433a-97ae-298ac0930e6a)

Reply:
> Nice! Might be cool to also have the `std::any::type_name` (or is there something better?) in the metadata file. Then, we could report the module path of the types as well.

That'd also be pretty helpful! I'll work on it right now.

Reply:
![image](https://github.com/Aleph-Alpha/ts-rs/assets/131818645/e67535c8-f803-475d-8938-dda60494ae30)

Reply:
That looks great, awesome!

Reply:
Should we add a big error message at the end saying something like "Your index.ts file is not valid due to these naming collisions" or something or just let the user find out from `tsc`?

Reply:
I think so! I dont see how a user could do anything useful with the invalid index file. Probably makes sense to not even write it to a file in that case.

Reply:
![image](https://github.com/Aleph-Alpha/ts-rs/assets/131818645/95cb1471-ba41-4494-b113-c8f013af154f)